pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -641,9 +641,12 @@ func (s *DiscoveryServer) initConnectionNode(discReq *xdsapi.DiscoveryRequest, c
 	}
 	con.mu.Unlock()
 
+	if err := con.modelNode.SetServiceInstances(s.Env.PushContext); err != nil {
+		return err
+	}
 	// Set the sidecarScope associated with this proxy if its a sidecar.
 	if con.modelNode.Type == model.SidecarProxy {
-		s.globalPushContext().SetSidecarScope(con.modelNode)
+		con.modelNode.SetSidecarScope(s.Env.PushContext)
```

Comment:
Maybe use the `s.globalPushContext()` to get better protection?

Reply:
make sense